technical-reports/format/design-token.md

```diff
@@ -130,6 +130,36 @@ For example:
 
 </aside>
 
+## Private
+
+Marks a token as private.
+
+Private tokens can be used to create aliases and other token combinations, reducing duplication and simplifying token outputs. Tools can resolve private tokens internally, but should not look to make them publicly available by default.
```

Comment:
This effectively makes this a write-only storage.

Once a designer marks a token as private it would become invisible to them because their tool would need to honour the flag.

It might be more interesting to have "filter" tools that teams can use to convert one or more "source" token files into a "filtered" final token file.

Reply:
🤔 Tools could be required to have a setting to show or hide `private` tokens.

Reply:
It depends on how you see these files being used. A token editor tool, could consume the input JSON and would be able to see all tokens and allow you to see/modify the private flag.

A translator on the other hand, would resolve private tokens, but would not export them to any output files (e.g. CSS, XML, JSON etc.)

I feel designers and developers should be consuming a processed output and see the input JSON as the source of truth, which only a certain set of designers modify. This could be an alternative approach to filter tools.

Reply:
So this wording would need to change to be clearer around different types of tooling.

Reply:
Would it then not be better to have a token management app/service which stores this information under `$extension`.

I don't think this requires a spec change.

--------

> which only a certain set of designers modify.

I don't think it makes sense to encode a hierarchy of control over tokens in the spec.

I do think it can be useful to have a filter to reduce clutter.

Reply:
> I don't think it makes sense to encode a hierarchy of control over tokens in the spec. (what if tokens are created by developers with dev tools in a certain org?)

Yeah I'm not saying this should be defined in the spec, it needs to be flexible for anyone to use. A token change could have disarsterous effects, though. So, this would no doubt be limited to certain editors who are working on large projects. I agree with you, the spec should definitely stay clear of all of this, I was only giving an example. My main point is that I feel designers and developers should consume the same data, albeit in different formats. Token editors, whoever they may be, would edit the input JSON - either directly, or via other tooling, and they would be able to see all data, including the visibility of a token.

Adding private token support, allows structuring and reusablity of internal values - it effectively makes things a lot cleaner for token editors. Further filtering could be beneficial for other teams, however, I see private tokens as a separate thing, which is beneficial to the spec. That's if, we have agreement on both designers and developers consuming the same (processed) data?

Reply:
My main hesitation with this is that I don't think `token editors` can ever be a well defined role with tools exclusively for them.

I don't think you can write a spec for `$private` because it's not realistic to define which tools should show or hide the tokens.

Ultimately anyone with a plain text editor can still open the tokens file and remove the `$private` flag.

I just don't think this is the right API/abstraction.
But I am just one person, with one persons point of view :)

Reply:
Writing a spec is extremely difficult and you need to be hesitant with any new request - I fully support that! :)

I guess I'm trying to understand how you would see all of this being used in the real would. Taking 'token editors' out of the equation, would you see the `tokens.json` as input and translated files as output? If so, then `$private` should make sense, as the properties configure the output. If not, I would like to know more about how and where this spec would be implemented.

FYI: I'm fully up for being more involved around all of this.

Reply:
I'm a community contributor and translation tool author (not an editor), so my opinions are my own and my concerns are influenced by my role as a translation tool author.

Personally I prefer to trust and empower downstream consumers.
Not hiding anything creates transparency which will foster trust and open communication between teams and team members.

-------

> Adds support for private tokens, to make defining tokens easier, without bloating any token outputs.

> would you see the tokens.json as input and translated files as output? If so, then $private should make sense, as the properties configure the output. If not, I would like to know more about how and where this spec would be implemented.

Is the issue you are trying to address that translation tools mostly generate code and do this for all tokens.

A tokens file following the design tokens methodology will have a lot of tokens intended for aliasing.
A translation tool for CSS will generate `--foo: #000;` declarations for each without.

The end result has a lot of bloat because the translation tool doesn't validate if a token value is used.

Reply:
Also see https://github.com/design-tokens/community-group/issues/110

Reply:
The issue that I'm facing is that I want to use **_tokens_** to create **tokens**. However, I don't want the **_tokens_** I used to create the **tokens** to be pomoted for use - They are used for computing other tokens, and I want to pass this information to translators so that they can choose how to handle them. An option could be passed to a translator to include them in the output. It's a way to structure `tokens.json` and inform translators which tokens are relevant and should be exported by default. It may only be a single token that is marked as private, but is used to create many other token values.

The example in the description is probably not the best. It hides colors from designers, which is not ideal.

Reply:
> Double checking because private has a very strong association with access control and that is a different thing.

`$hidden`?

Reply:
Classes in certain programming languages also have to concepts of `abstract` and `final`. I think that has some overlap.

`abstract` -> only intended to be used to construct other tokens
`final` -> never intended to be used to construct other tokens

Tokens with `$abstract` :
- can only be used via an alias
- never possible to access their value directly

Tokens with `$final` :
- can never be aliased

_You don't need both concepts._

Reply:
After thinking about this a bit more, I don't think we need to go that far, we're not designing a new programming language here - it's essentially a configuration file. `$abstract` and `$final` feel a bit too much.

I agree that `$private` is the wrong name, however, and I would suggest it to be more like a visibility state, hidden/visible. Token editing tools would be able to see all tokens and modify the visibility state for translators.

`$hidden` seems like a good option.